src/libraries/Microsoft.PowerFx.Core/Syntax/NamedFormulas.cs

```diff
@@ -0,0 +1,116 @@
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Syntax.Nodes;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Syntax
+{
+    /// <summary>
+    /// This encapsulates a named formula: its original script, the parsed result, and any parse errors.
+    /// </summary>
+    internal class NamedFormulas
+    {
+        /// <summary>
+        /// A script containing one or more named formulas.
+        /// </summary>
+        public readonly string Script;
+
+        // The language settings used for parsing this script.
+        // May be null if the script is to be parsed in the current locale.
+        public readonly ILanguageSettings Loc;
+
+        public Dictionary<DName, TexlNode> FormulasResult;
+
+        public bool IsParsed => FormulasResult != null;
+
+        public bool HasParseErrors { get; private set; }
+
+        private List<TexlError> _errors;
+
+        private List<Formula> _formulas;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NamedFormulas"/> class.
+        /// </summary>
+        /// <param name="script"></param>
+        /// <param name="loc"></param>
+        public NamedFormulas(string script, ILanguageSettings loc = null)
+        {
+            Contracts.AssertValue(script);
+            Contracts.AssertValueOrNull(loc);
+
+            Script = script;
+            Loc = loc;
+        }
+
+        /// <summary>
+        /// Ensures that the named formulas have been parsed and if not, parses them.
+        /// </summary>
+        /// <returns></returns>
+        public bool EnsureParsed()
+        {
+            if (FormulasResult == null)
+            {
+                Contracts.AssertValue(Script);
+                Contracts.AssertValueOrNull(Loc);
+                var result = TexlParser.ParseFormulasScript(Script, loc: Loc);
+                FormulasResult = result.NamedFormulas;
+                _errors = result.Errors;
+                HasParseErrors = result.HasError;
+                Contracts.AssertValue(FormulasResult);
+            }
+
+            return _errors == null;
+        }
+
+        /// <summary>
+        /// Returns any parse errors.
+        /// </summary>
+        /// <returns></returns>
+        public IEnumerable<TexlError> GetParseErrors()
+        {
+            Contracts.AssertValue(Script);
+            Contracts.Assert(IsParsed, "Should call EnsureParsed() first!");
+            return _errors ?? Enumerable.Empty<TexlError>();
+        }
+
+        /// <summary>
+        /// Returns a Formula object for each named formula.
+        /// </summary>
+        /// <returns></returns>
+        public List<Formula> GetFormulas()
+        {
+            _formulas = new List<Formula>();
+            if (FormulasResult != null)
+            {
+                foreach (var kvp in FormulasResult)
+                {
+                    TryGetSubscript(kvp.Key, out var subScript);
```

Comment:
Seems odd to call a tryget that just looks up into the dict we're iterating over, and not check the result. Could just be 
`var subscript = kvp.Value.GetCompleteSpan.GetFragment(Script);`